palette/moreland/convert.go

```diff
@@ -0,0 +1,229 @@
+		A: 500.0 * (tempX - tempY),
+		B: 200 * (tempY - tempZ),
+	}
+}
+
+// xyz2labHelper is an intermediate step in converting from CIE XYZ to CIE LAB.
+func xyz2labHelper(v float64) float64 {
+	if v > 0.008856 {
+		return math.Pow(v, 1.0/3.0)
+	}
+	return 7.787*v + 16.0/116.0
+}
+
+// sRGBA represents a color within the sRGB color space, with an alpha channel
+// but not premultiplied.
+type sRGBA struct {
+	R, G, B, A float64
+}
+
+// rgb converts the receiver to linear RGB space.
+func (c sRGBA) rgb() rgb {
+	return rgb{
+		R: sToLinear(c.R),
+		G: sToLinear(c.G),
+		B: sToLinear(c.B),
+	}
+}
+
+// sToLinear converts from an sRGB component to a linear RGB component.
+func sToLinear(v float64) float64 {
+	if v > 0.04045 {
+		return math.Pow((v+0.055)/1.055, 2.4)
+	}
+	return v / 12.92
+}
+
+// RGBA implements the color.Color interface.
+func (c sRGBA) RGBA() (r, g, b, a uint32) {
+	return uint32(c.R * c.A * 65535.0), uint32(c.G * c.A * 65535.0), uint32(c.B * c.A * 65535.0), uint32(c.A * 65535.0)
+}
+
+// cieLAB converts a sRGB color to a CIE LAB color.
+func (c sRGBA) cieLAB() cieLAB {
+	return c.rgb().cieXYZ().cieLAB()
+}
+
+// colorTosRGBA converts a color to an sRGBA.
+func colorTosRGBA(c color.Color) sRGBA {
+	r, g, b, a := c.RGBA()
+	if a == 0 {
+		return sRGBA{}
+	}
+	return sRGBA{
+		R: float64(r) / float64(a),
+		G: float64(g) / float64(a),
+		B: float64(b) / float64(a),
+		A: float64(a) / 65535.0,
+	}
+}
+
+// clamp forces all channels in c to be within the range (0,1).
+func (c *sRGBA) clamp() {
+	if c.R > 1 {
+		c.R = 1
+	}
+	if c.G > 1 {
+		c.G = 1
+	}
+	if c.B > 1 {
+		c.B = 1
+	}
+	if c.A > 1 {
+		c.A = 1
+	}
+	if c.R < 0 {
+		c.R = 0
+	}
+	if c.G < 0 {
+		c.G = 0
+	}
+	if c.B < 0 {
+		c.B = 0
+	}
+	if c.A < 0 {
+		c.A = 0
+	}
+}
+
+// cieLAB represents a color in CIE LAB space.
+type cieLAB struct {
+	L, A, B float64
+}
+
+// sRGB converts a CIE LAB color to an sRGBA color, where alpha is opacity
+// between 0 and 1.
+func (c cieLAB) sRGBA(alpha float64) sRGBA {
+	return c.cieXYZ().rgb().sRGBA(alpha)
+}
+
+// cieXYZ converts a CIE LAB color to a CIE XYZ color.
+func (c cieLAB) cieXYZ() cieXYZ {
+	// Reference white-point D65
+	const xn, yn, zn = 0.95047, 1.0, 1.08883
+	return cieXYZ{
+		X: xn * lab2xyzHelper((c.A/500)+(c.L+16)/116),
+		Y: yn * lab2xyzHelper((c.L+16.)/116.),
+		Z: zn * lab2xyzHelper((c.L+16.)/116.-(c.B/200.)),
+	}
+}
+
+// lab2xyzHelper is an intermediate step in converting from CIE LAB to CIE XYZ.
+func lab2xyzHelper(v float64) float64 {
+	const (
+		xlim = 0.008856
+		a    = 7.787
+		b    = 16. / 116.
+		ylim = a*xlim + b
+	)
+	if v > ylim {
+		return v * v * v
+	}
+	return (v - b) / a
+}
+
+// MSH converts a cieLAB color to MSH space.
+func (c cieLAB) MSH() msh {
+	m := math.Pow(c.L*c.L+c.A*c.A+c.B*c.B, 0.5)
+	return msh{
+		M: m,
+		S: math.Acos(c.L / m),
+		H: math.Atan2(c.B, c.A),
+	}
+}
+
+// MSH defines a color in Magnitude-Saturation-Hue color space.
+type msh struct {
+	M, S, H float64
```

Comment:
From the equations used to convert to this space, and the fact that CIE LAB colors don't seem to have rigidly defined ranges for the A and B values, it appears that there may not be well-defined ranges for any of the three components.